src/common/libjob/jobspec1.h

```diff
@@ -0,0 +1,140 @@
+ */
+int flux_jobspec1_env_set (flux_jobspec1_t *jobspec,
+                           const char *name,
+                           const char *value,
+                           int overwrite);
+
+/* Change or add an entry to a Jobspec's environment.
+ * 'entry' should be a string of the form 'name=value'. Unlike 'putenv(3)',
+ * the string will be copied.
+ * Return 0 on success, nonzero with errno set on error.
+ */
+int flux_jobspec1_env_put (flux_jobspec1_t *jobspec, const char *entry);
+
+/* Change or add multiple entries to a Jobspec's environment.
+ * 'env' should be an 'environ(7)'-style array, where NULL indicates an
+ * empty environment (and in which case this function has no effect).
+ * Return 0 on success, -1 with errno set on error.
+ */
+int flux_jobspec1_env_putmany (flux_jobspec1_t *jobspec, char **env);
+
+/* Direct the stdin of a jobspec to a path given by
+ * 'stdin'. Return 0 on success, -1 on error.
+ */
+int flux_jobspec1_set_stdin (flux_jobspec1_t *jobspec, const char *path);
+
+/* Direct the stdout of a jobspec to a path given by
+ * 'stdout'. Return 0 on success, -1 on error.
+ */
+int flux_jobspec1_set_stdout (flux_jobspec1_t *jobspec, const char *path);
+
+/* Direct the stderr of a jobspec to a path given by
+ * 'stderr'. Return 0 on success, -1 on error.
+ */
+int flux_jobspec1_set_stderr (flux_jobspec1_t *jobspec, const char *path);
+
+/* Set the working directory of a jobspec.
+ * Return 0 on success, -1 on error.
+ */
+int flux_jobspec1_set_cwd (flux_jobspec1_t *jobspec, const char *cwd);
+
+/* Encode a jobspec to a string, e.g. for usage with ``flux_job_submit()``.
+ * 'flags' should be 0.
+ * Return NULL with errno set on error.
+ * The return value must be released with ``free()``.
+ */
+char *flux_jobspec1_encode (flux_jobspec1_t *jobspec, size_t flags);
+
+/* Create and return a minimum viable V1 Jobspec.
+ * The jobspec will have stdin, stdout, stderr all directed to the KVS,
+ * and the environment will be empty.
+ * The jobspec must be released with 'flux_jobspec1_free'.
+ * 'argc' and 'argv' should collectively define the command and its
+ * arguments.
+ * 'tasks' should be the number of tasks to launch
+ * 'cores_per_task' should be the number of cores per task to allocate
+ * 'gpus_per_task' should be the number of gpus per task to allocate
+ * 'nodes' should be the number of nodes to spread the allocated cores
+ * and gpus across. If 0, the scheduler will determine how to distribute
+ * the resources.
+ * Return NULL on error with errno set.
+ */
+flux_jobspec1_t *flux_jobspec1_from_command (int argc,
+                                             char **argv,
+                                             int ntasks,
+                                             int cores_per_task,
+                                             int gpus_per_task,
+                                             int nnodes,
+                                             double duration);
```

Comment:
I wonder if we should add an optional `char **environ` to this constructor. e.g.
```suggestion
flux_jobspec1_t *flux_jobspec1_from_command (int argc,
                                             char **argv,
                                             char **environ,
                                             int ntasks,
                                             int cores_per_task,
                                             int gpus_per_task,
                                             int nnodes,
                                             double duration);
```

If `environ == NULL` then the function would work as it does now and create a jobspec with an empty environment. Otherwise, the `environ` would be copied into the jobspec.

Since a common use case might be to copy the current environment, this would allow many users of the API to use this one call instead two to create simple jobspec. It would also remove the requirement for `flux_jobspec1_env_putmany()` in the API.